### ISSUE TITLE
Remote reconciliation strategy

### ISSUE THREAD

**Issue by Tomasz-Smelcerz-SAP** [kyma-project/lifecycle-manager#30]
The purpose of this issue is to define how do we want to architecture and implement the remote reconcilitation.

The remote reconciliation assumes:
- Component operators are running in the Control Plane
- Component operator CRs are created in the Control Plane
- Running instances of component operators are reconciling Kyma components in remote clusters (customer clusters)

The challenges found initially:
- Scaling of the solution. Consider one hundred clusters are to be upgraded starting Tuesday morning. How long does it take overall, how long does it take for a single cluster on average and what's the latency in CR status updates?
- Do we want to control the ordering of upgrades between clusters? See below for explanation.

1. Scaling of the solution
    The operators can't scale. There's always only one instance of an operator. How do we reconcile 100 clusters in parallel using just a single golang process? 
    - goroutines won't work - do you have a Node performant enough?
    - horizontal scaling won't work - you can't increase replicas for an operator
    - what about status update latency? We should update status on CR's to something like "Pending" ASAP, for every CR, regardless of the time it takes to actually upgrade the related clusters. It means the actual workload should somehow be scheduled to the "background" or to some external entity, and the operator should process CR's without any substantial delay.
    
    Possible solutions:
    - just measure what we can achieve right now, if it's not enough, try to find a solution :)
    

3. Ordering of upgrades problem - latency, resources.

    Assume we have just three component operators: A, B, C.
Let's also assume all component operators performs the reconciliation step in a similar time.
Assume we have 10 clusters to upgrade.
We can easily show that the order of processing between component operators does change the observed latency and resource allocation.

    First model (worst-case):

    | Steps in time | comp. A is processing | comp. B  is processing | comp. C is processing | clusters ready after the step |
    | - | --------- | --------- | ----------| - |
    | 1 |  cluster 1 |  cluster 1 | cluster 10 | - |
    | 2 | cluster 2 |  cluster 2 | cluster 9 | - |
    | 3 |  cluster 3 | cluster 3 | cluster 8 | - |
    | 4 |  cluster 4 | cluster 4 | cluster 7 | - |
    | 5 |  cluster 5 | cluster 10 | cluster 6 | - |
    | 6 |  cluster 6 | cluster 9 | cluster 5 | - |
    | 7 |  cluster 7 | cluster 8 | cluster 4 | 4 |
    | 8 |  cluster 8 | cluster 7 | cluster 3 | 4, 8, 7, 3 |
    | 9 |  cluster 9 | cluster 6 | cluster 2 | 4, 8, 7, 3, 9, 6, 2|
    | 10 |  cluster 10 | cluster 5 | cluster 1 | 4, 8, 7, 3, 9, 6, 2, 10, 5, 1|

    In this model, we are actively reconciling up to 6 clusters at any time, and the average latency for a cluster to be ready is 8.8 (in the artificial "Steps in time" units)

    Second model (best-case):

    | Steps in time | comp. A is processing | comp. B  is processing | comp. C is processing | clusters ready after the step|
    | - | --------- | --------- | ----------| - |
    | 1 |  cluster 1 |  cluster 1 | cluster 1 | 1 |
    | 2 | cluster 2 |  cluster 2 | cluster 2 | 1, 2 |
    | 3 |  cluster 3 | cluster 3 | cluster 3 | 1, 2, 3 |
    | 4 |  cluster 4 | cluster 4 | cluster 4 | 1, 2, 3, 4 |
    | 5 |  cluster 5 | cluster 5 | cluster 5 | 1, 2, 3, 4, 5|
    | 6 |  cluster 6 | cluster 6 | cluster 6 | 1, 2, 3, 4, 5, 6 |
    | 7 |  cluster 7 | cluster 7 | cluster 7 | 1, 2, 3, 4, 5, 6, 7 |
    | 8 |  cluster 8 | cluster 8 | cluster 8 | 1, 2, 3, 4, 5, 6, 7, 8 |
    | 9 |  cluster 9 | cluster 9 | cluster 9 | 1, 2, 3, 4, 5, 6, 7, 8, 9 |
    | 10 |  cluster 10 | cluster 10 | cluster 10 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 |

    In this model, we are actively reconciling only 1 cluster at a time, and the average latency for a cluster to be ready is 5.5 (in the artificial "Steps in time" units).

    Notice that the amount of work to be done is the same in both models, just the ordering makes the difference.
    Our current proposed strategy is: Every component operator is independent and is processing clusters in it's own order/pace
    This unfortunately leads more towards the first execution model, which is not optimal.

    Another thing is resource consumption. If we actively reconcile 6 different clusters at a time, we need 6 different k8s golang clients, 6 HTTPS streams, and so on.
    Finding a way to share such resources between operators (a common external worker, for example), along with processing clusters in some synchronized order, would reduce resource consumption along with the latency.

    As always, the benefits from the possible improvements here may not justify the effort - it should be measured.

**Comment by Tomasz-Smelcerz-SAP** [comment]
No longer relevant.
